mllib/src/test/scala/org/apache/spark/ml/clustering/KMeansSuite.scala

```diff
@@ -137,18 +142,53 @@ class KMeansSuite extends SparkFunSuite with MLlibTestSparkContext with DefaultR
       assert(model.clusterCenters === model2.clusterCenters)
     }
     val kmeans = new KMeans()
-    testEstimatorAndModelReadWrite(kmeans, dataset, KMeansSuite.allParamSettings, checkModelData)
+    testEstimatorAndModelReadWrite(kmeans, dataset, KMeansSuite.allParamSettings, checkModelData,
+      Map("initialModel" -> (checkModelData _).asInstanceOf[(Any, Any) => Unit]))
+  }
+
+  test("Initialize using given cluster centers") {
+    val kmeans = new KMeans().setK(k).setSeed(1).setMaxIter(1)
+    val oneIterModel = kmeans.fit(dataset)
+    val twoIterModel = kmeans.copy(ParamMap(ParamPair(kmeans.maxIter, 2))).fit(dataset)
+    val oneMoreIterModel = kmeans.setInitialModel(oneIterModel).fit(dataset)
+
+    twoIterModel.clusterCenters.zip(oneMoreIterModel.clusterCenters)
+      .foreach { case (center1, center2) => assert(center1 ~== center2 absTol 1E-8) }
+  }
+
+  test("Initialize using wrong model") {
+    val kmeans = new KMeans().setK(k).setSeed(1).setMaxIter(10)
+
+    val wrongKModel = KMeansSuite.generateRandomKMeansModel(3, k + 1)
+    val wrongKModelThrown = intercept[IllegalArgumentException] {
+      kmeans.setInitialModel(wrongKModel).fit(dataset)
+    }
+    assert(wrongKModelThrown.getMessage.contains("mismatched cluster count"))
+
+    val wrongDimModel = KMeansSuite.generateRandomKMeansModel(4, k)
+    val wrongDimModelThrown = intercept[IllegalArgumentException] {
+      kmeans.setInitialModel(wrongDimModel).fit(dataset)
+    }
+    assert(wrongDimModelThrown.getMessage.contains("mismatched dimension"))
   }
 }
 
```

Comment:
Let's have a test case that without explicitly setting k, k can be inferred from initial model.